scripts/triggers-end-to-end-tests/functions/index.js

```diff
@@ -0,0 +1,78 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+
+/*
+ * Log snippets that the driver program above checks for. Be sure to update
+ * ../test.js if you plan on changing these.
+ */
+const RTDB_FUNCTION_LOG = '========== RTDB FUNCTION ==========';
+const FIRESTORE_FUNCTION_LOG = '========== FIRESTORE FUNCTION ==========';
+
+/*
+ * We install onWrite triggers for START_DOCUMENT_NAME in both the firestore and
+ * database emulators. From each respective onWrite trigger, we write a document
+ * to both the firestore and database emulators. This exercises the
+ * bidirectional communication between cloud functions and each emulator.
+ */
+const START_DOCUMENT_NAME = "test/start";
+const END_DOCUMENT_NAME = "test/done";
+
+admin.initializeApp();
+
+exports.deleteFromFirestore = functions.https.onRequest(async (req, res) => {
+    await admin.firestore().doc(START_DOCUMENT_NAME).delete();
+    res.json({ deleted: true });
+});
+
+exports.deleteFromRtdb = functions.https.onRequest(async (req, res) => {
+    await admin.database().ref(START_DOCUMENT_NAME).remove();
+    res.json({ deleted: true });
+});
+
+exports.writeToFirestore = functions.https.onRequest(async (req, res) => {
+    const ref = admin.firestore().doc(START_DOCUMENT_NAME);
+    await ref.set({ start: (new Date().toISOString()) });
+    ref.get().then((snap) => {
+        res.json({ data: snap.data() });
+    });
+});
+
+exports.writeToRtdb = functions.https.onRequest(async (req, res) => {
+    const ref = admin.database().ref(START_DOCUMENT_NAME);
+    await ref.set({ start: (new Date().toISOString()) });
+    ref.once('value', (snap) => {
+        res.json({ data: snap });
+    });
+});
+
+exports.firestoreReaction = functions.firestore.document(START_DOCUMENT_NAME).onWrite(async (change, ctx) => {
+    console.log(FIRESTORE_FUNCTION_LOG);
+    /*
+     * Write back a completion timestamp to the firestore emulator. The test
+     * driver program checks for this by querying the firestore emulator
+     * directly.
+     */
+    const ref = admin.firestore().doc(END_DOCUMENT_NAME + "_from_firestore");
```

Comment:
What's the purpose of the `_from_firestore` and `_from_database` suffixes?

Reply:
Ohhh I see we're writing to RTDB from Firestore functions (and vice-versa).

Does that test any particular concern?

Reply:
The suffixes indicate whether the object was written from a firestore or realtime database trigger.

Reply:
Is there any reason to suspect that writing to RTDB from a Firestore function would be different than writing to RTDB from a RTDB function?

IMO it would be sufficient just to test RTDB<-->Functions<-->RTDB and Firestore<-->Functions<-->Firestore

Reply:
Nothing specific, but the key change we missed in the original PR was having the functions emulator configure the "in-cloud-function" admin sdk to point at the database emulator. I thought ensuring this worked for both emulators was a valuable exercise. I think covering this specific use-case in our tests might also be valuable for developers looking to test a firestore -> database or database -> firestore migration/sync pipelines.

Reply:
SGTM since you already wrote it and I can see no harm in this.